### PR TITLE
[Adding] ability to revoke user by nkey only

### LINKED ISSUE #302: Allow revoking users that are not in the nkey store
Currently `nsc` prevents revoking a user that is not present in the filesystem

```sh
nsc revocations delete_user -a KUBECON -n UBS7HNAO2NPNLT...
Error: user UBS7HNAO2NPNLT3DFXH2ABAJT55N2AF67UANPL7S7AFNKYMNSMBTZDMX does not exist in the current account
```

### PR DESCRIPTION
**Pull request by matthiashanel**
Fixes #302

Signed-off-by: Matthias Hanel <mh@synadia.com>

### CODE REVIEW

File: cmd/revokeuser.go [hunk 1]
```diff
@@ -40,6 +40,7 @@ func createRevokeUserCmd() *cobra.Command {
 	}
 	cmd.Flags().StringVarP(&params.user, "name", "n", "", "user name")
 	cmd.Flags().IntVarP(&params.at, "at", "", 0, "revokes all user credentials created before a Unix timestamp ('0' is treated as now)")
+	cmd.Flags().StringVarP(&params.userPubKey, "public-user-key", "u", "", "public user key (use when user was generated outside of nsc)")
```
Comment (aricart) [inline cmd/revokeuser.go:44]: --user-public-key is prob the right flag name

File: cmd/revokeuser.go [hunk 2]
```diff
@@ -110,33 +111,38 @@ func (p *RevokeUserParams) Load(ctx ActionCtx) error {
...
 		n := ctx.StoreCtx().DefaultUser(p.AccountContextParams.Name)
 		if n != nil {
 			p.user = *n
 		}
 	}
 
-	if p.user == "" {
-		return fmt.Errorf("user is required")
+	if p.userPubKey == "" && p.user == "" {
+		return fmt.Errorf("user or public-user-key are required")
```
Comment (aricart) [inline cmd/revokeuser.go:122]: same here

File: cmd/revokeuser.go [hunk 3]
```diff
@@ -110,33 +111,38 @@ func (p *RevokeUserParams) Load(ctx ActionCtx) error {
...
+	} else if p.userPubKey != "" && p.user != "" {
+		return fmt.Errorf("user and public-user-key are mutually exclusive")
```
Comment (aricart) [inline cmd/revokeuser.go:124]: and here.

File: cmd/revokeuser.go [hunk 4]
```diff
@@ -110,33 +111,38 @@ func (p *RevokeUserParams) Load(ctx ActionCtx) error {
...
-	}
+		if err != nil || userClaim == nil {
+			return fmt.Errorf("user is required")
+		}
 
-	if err != nil || userClaim == nil {
-		return fmt.Errorf("user is required")
+		p.userPubKey = userClaim.Subject
+	} else if !nkeys.IsValidPublicUserKey(p.userPubKey) {
+		return fmt.Errorf("public-user-key %q is not a valid public user nkey", p.userPubKey)
```
Comment (aricart) [inline cmd/revokeuser.go:143]: public-user-key

File: cmd/revokeuser.go [hunk 5]
```diff
@@ -110,33 +111,38 @@ func (p *RevokeUserParams) Load(ctx ActionCtx) error {
...
-		return err
-	}
+		if err != nil || userClaim == nil {
+			return fmt.Errorf("user is required")
+		}
 
-	if err != nil || userClaim == nil {
-		return fmt.Errorf("user is required")
+		p.userPubKey = userClaim.Subject
+	} else if !nkeys.IsValidPublicUserKey(p.userPubKey) {
```
Comment (aricart) [inline cmd/revokeuser.go:142]: Key validation should happen on the validate, that way the interactive doesn't have to validate as well

Comment (matthiashanel) [reply cmd/revokeuser.go]: fixed

**Comment by matthiashanel** [comment]
Added support to clear revocations by nkey.

Changed the interactive mode for clear to only show revoked nkeys (vs. all users)
When available the user name is presented too.

Also aliased commands to work with -

```
 16:22:52  ✘ 1  ~/repos/nsc  revoke-by-pub *13 ?1  nsc revocations delete-user -u UCLMBZ5CBDRDG2TAYOJK23U7IGKPTTW7DTWNOP4TUW4PAB3GRUSKXG3N
[ OK ] cleared user revocation for account UCLMBZ5CBDRDG2TAYOJK23U7IGKPTTW7DTWNOP4TUW4PAB3GRUSKXG3N
Cleared revocation of user  with public key UCLMBZ5CBDRDG2TAYOJK23U7IGKPTTW7DTWNOP4TUW4PAB3GRUSKXG3N
 16:23:02  ~/repos/nsc  revoke-by-pub *13 ?1  nsc revocations list_users
╭──────────────────────────────────────────────────────────────────────────────────────────╮
│                                 Revoked Users for testme                                 │
├──────────────────────────────────────────────────────────┬───────────────────────────────┤
│ Public Key                                               │ Revoke Credentials Before     │
├──────────────────────────────────────────────────────────┼───────────────────────────────┤
│ UAZZM3O3EP2EGGQFCBZPIXEKGNXTHT6USNM3BXMQFVX5M65TQCNN3H24 │ Fri, 02 Oct 2020 15:18:20 EDT │
╰──────────────────────────────────────────────────────────┴───────────────────────────────╯
 16:23:16  ~/repos/nsc  revoke-by-pub *13 ?1  nsc revocations delete-user -u UAZZM3O3EP2EGGQFCBZPIXEKGNXTHT6USNM3BXMQFVX5M65TQCNN3H24
[ OK ] cleared user revocation for account UAZZM3O3EP2EGGQFCBZPIXEKGNXTHT6USNM3BXMQFVX5M65TQCNN3H24
Cleared revocation of user  with public key UAZZM3O3EP2EGGQFCBZPIXEKGNXTHT6USNM3BXMQFVX5M65TQCNN3H24
 16:23:18  ~/repos/nsc  revoke-by-pub *13 ?1  nsc revocations list_users
╭────────────────────────────────────────╮
│        Revoked Users for testme        │
├────────────┬───────────────────────────┤
│ Public Key │ Revoke Credentials Before │
├────────────┼───────────────────────────┤
╰────────────┴───────────────────────────╯
 16:23:19  ~/repos/nsc  revoke-by-pub *13 ?1 
```

File: cmd/revokeclearuser.go [hunk 6]
```diff
@@ -74,11 +79,43 @@ func (p *ClearRevokeUserParams) PreInteractive(ctx ActionCtx) error {
...
 		return err
 	}
-	if p.user == "" {
-		p.user, err = ctx.StoreCtx().PickUser(p.AccountContextParams.Name)
-		if err != nil {
-			return err
+
+	st := ctx.StoreCtx().Store
+	accClaim, err := st.ReadAccountClaim(p.AccountContextParams.Name)
+	if err != nil {
```
Comment (aricart) [inline cmd/revokeclearuser.go:85]: if err != nil || len(accClaim.Revocations) == 0 {
   return err
}

File: cmd/revokeclearuser.go [hunk 7]
```diff
@@ -74,11 +79,43 @@ func (p *ClearRevokeUserParams) PreInteractive(ctx ActionCtx) error {
...
+		return err
+	} else if len(accClaim.Revocations) == 0 {
+		return nil
+	}
+	keyToUserName := map[string]string{}
+	if users, err := st.ListEntries(store.Accounts, p.AccountContextParams.Name, store.Users); err != nil {
```
Comment (aricart) [inline cmd/revokeclearuser.go:89]: to me this reads better:
users, err := ...
if err != nil {
  return err
}

// no else required

File: cmd/revokeclearuser.go (hunk 6 above)
Comment (matthiashanel) [reply cmd/revokeclearuser.go]: fixed

File: cmd/revokeclearuser.go [hunk 8]
```diff
@@ -93,22 +130,26 @@ func (p *ClearRevokeUserParams) Load(ctx ActionCtx) error {
...
 		if n != nil {
 			p.user = *n
 		}
 	}
-
-	if p.user == "" {
-		return fmt.Errorf("user is required")
+	if p.userPubKey == "" && p.user == "" {
+		return fmt.Errorf("user or user-public-key are required")
+	} else if p.userPubKey != "" && p.user != "" {
```
Comment (aricart) [inline cmd/revokeclearuser.go:139]: there's no else...

File: cmd/revokeuser.go [hunk 9]
```diff
@@ -75,20 +77,24 @@ func (p *RevokeUserParams) canParse(s string) error {
...
-
-	if err = p.AccountContextParams.Edit(ctx); err != nil {
+	if err := p.AccountContextParams.Edit(ctx); err != nil {
 		return err
 	}
-	if p.user == "" {
-		p.user, err = ctx.StoreCtx().PickUser(p.AccountContextParams.Name)
-		if err != nil {
-			return err
+	if byName, err := cli.Confirm("Do you want to revoke a user (known to nsc) by name?", true); err != nil {
```
Comment (aricart) [inline cmd/revokeuser.go:83]: byName, err := ...
if err != nil { 
  return...
}

File: cmd/revokeuser.go [hunk 10]
```diff
@@ -110,22 +117,25 @@ func (p *RevokeUserParams) Load(ctx ActionCtx) error {
 		return err
 	}
 
-	if p.user == "" {
+	if p.user == "" && p.userPubKey == "" {
 		n := ctx.StoreCtx().DefaultUser(p.AccountContextParams.Name)
 		if n != nil {
 			p.user = *n
 		}
```
Comment (aricart) [inline cmd/revokeuser.go:124]: if n == nil {
   fmt.Errorf("some error)
}
p.user = *n

File: cmd/revokeuser.go [hunk 11]
```diff
@@ -110,22 +117,25 @@ func (p *RevokeUserParams) Load(ctx ActionCtx) error {
...
 	}
 
-	if p.user == "" {
-		return fmt.Errorf("user is required")
+	if p.userPubKey == "" && p.user == "" {
+		return fmt.Errorf("user or user-public-key are required")
+	} else if p.userPubKey != "" && p.user != "" {
```
Comment (aricart) [inline cmd/revokeuser.go:129]: this won't happen with the above...

**Review by aricart — APPROVED** [review]
the if/if else/else type stuff gets really dense to look at - I think you should simplify it.

File: cmd/revokeuser.go (hunk 11 above)
Comment (matthiashanel) [reply cmd/revokeuser.go]: it will, if the above does not set a user

**Pull request merged**